File: backend/src/main/java/com/meditrack/auth/AuthService.java
package com.meditrack.auth;

import com.meditrack.enums.AccountStatus;
import com.meditrack.enums.OtpType;
import com.meditrack.security.JwtService;
import com.meditrack.user.User;
import com.meditrack.user.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final OtpRequestRepository otpRequestRepository;
    private final OtpService otpService;
    private final JwtService jwtService;
    private final PasswordEncoder passwordEncoder;
    private final AuthMapper authMapper;

    public AuthResponseDTO register(RegisterRequestDTO dto) {
        if (userRepository.existsByEmail(dto.getEmail())) {
            throw new RuntimeException("Email already registered");
        }
        if (dto.getMobile() != null && userRepository.existsByMobile(dto.getMobile())) {
            throw new RuntimeException("Mobile already registered");
        }

        User user = User.builder()
                .firstName(dto.getFirstName())
                .lastName(dto.getLastName())
                .email(dto.getEmail())
                .mobile(dto.getMobile())
                .passwordHash(passwordEncoder.encode(dto.getPassword()))
                .build();

        userRepository.save(user);
        otpService.generateAndSendOtp(user.getId(), user.getEmail(), OtpType.EMAIL_VERIFY);

        return authMapper.toRegisterResponse(user);
    }

    public AuthResponseDTO verifyOtp(VerifyOtpDTO dto) {
        User user = userRepository.findById(dto.getUserId())
                .orElseThrow(() -> new RuntimeException("User not found"));

        otpService.verifyOtp(user.getId(), dto.getOtp(), OtpType.EMAIL_VERIFY);

        user.setEmailVerified(true);
        user.setAccountStatus(AccountStatus.ACTIVE);
        userRepository.save(user);

        return authMapper.toVerifyResponse(user);
    }

    public AuthResponseDTO login(LoginRequestDTO dto) {
        User user = userRepository
                .findByEmailOrMobile(dto.getUsername(), dto.getUsername())
                .orElseThrow(() -> new RuntimeException("Invalid credentials"));

        if (!passwordEncoder.matches(dto.getPassword(), user.getPasswordHash())) {
            throw new RuntimeException("Invalid credentials");
        }

        if (!user.getEmailVerified()) {
            throw new RuntimeException("Email not verified");
        }

        if (user.getAccountStatus() != AccountStatus.ACTIVE) {
            throw new RuntimeException("Account is " + user.getAccountStatus());
        }

        user.setLastLoginAt(LocalDateTime.now());
        userRepository.save(user);

        return authMapper.toAuthResponse(
                user,
                jwtService.generateAccessToken(user.getId()),
                jwtService.generateRefreshToken(user.getId())
        );
    }

    public AuthResponseDTO forgotPassword(ForgotPasswordDTO dto) {
        User user = userRepository
                .findByEmailOrMobile(dto.getUsername(), dto.getUsername())
                .orElseThrow(() -> new RuntimeException("User not found"));

        otpService.generateAndSendOtp(user.getId(), user.getEmail(), OtpType.PASSWORD_RESET);

        return AuthResponseDTO.builder()
                .message("OTP sent successfully")
                .otpExpiresIn(300)
                .build();
    }

    public AuthResponseDTO resetPassword(ResetPasswordDTO dto) {
        User user = userRepository
                .findByEmailOrMobile(dto.getUsername(), dto.getUsername())
                .orElseThrow(() -> new RuntimeException("User not found"));

        otpService.verifyOtp(user.getId(), dto.getOtp(), OtpType.PASSWORD_RESET);

        user.setPasswordHash(passwordEncoder.encode(dto.getNewPassword()));
        userRepository.save(user);

        return AuthResponseDTO.builder()
                .message("Password reset successfully")
                .build();
    }

    public AuthResponseDTO refreshToken(RefreshTokenDTO dto) {
        if (!jwtService.isTokenValid(dto.getRefreshToken())) {
            throw new RuntimeException("Invalid refresh token");
        }

        String userId = jwtService.extractUserId(dto.getRefreshToken());
        userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        return AuthResponseDTO.builder()
                .accessToken(jwtService.generateAccessToken(userId))
                .refreshToken(jwtService.generateRefreshToken(userId))
                .message("Token refreshed successfully")
                .build();
    }
}

File: frontend/src/services/authService.ts
// src/services/authService.ts

import apiClient from './apiClient';

export interface LoginRequest {
  username: string; // email OR mobile
  password: string;
}

export interface RegisterRequest {
  firstName: string;
  lastName: string;
  email: string;
  mobile: string;
  password: string;
}

export interface VerifyOtpRequest {
  userId: string;
  otp: string;
}

export interface ForgotPasswordRequest {
  username: string; // email OR mobile
}

export interface ResetPasswordRequest {
  username: string;
  otp: string;
  newPassword: string;
}

class AuthService {
  // REGISTER
 // REGISTER
async register(data: RegisterRequest) {
  try {
    const response = await apiClient.post(
      '/auth/register',
      data
    );

    console.log(
      'REGISTER SUCCESS =>',
      JSON.stringify(response.data, null, 2)
    );

    return response.data;
  } catch (error: any) {
    console.log(
      'REGISTER STATUS =>',
      error?.response?.status
    );

    console.log(
      'REGISTER ERROR DATA =>',
      JSON.stringify(
        error?.response?.data,
        null,
        2
      )
    );

    throw error;
  }
}

  // VERIFY OTP
  async verifyOtp(data: VerifyOtpRequest) {
    const response = await apiClient.post(
      '/auth/verify-otp',
      data
    );

    return response.data;
  }

  // LOGIN
  async login(data: LoginRequest) {
    const response = await apiClient.post(
      '/auth/login',
      data
    );

    return response.data;
  }

  // FORGOT PASSWORD
  async forgotPassword(
    data: ForgotPasswordRequest
  ) {
    const response = await apiClient.post(
      '/auth/forgot-password',
      data
    );

    return response.data;
  }

  // RESET PASSWORD
  async resetPassword(
    data: ResetPasswordRequest
  ) {
    const response = await apiClient.post(
      '/auth/reset-password',
      data
    );

    return response.data;
  }

  // REFRESH TOKEN
  async refreshToken(refreshToken: string) {
    const response = await apiClient.post(
      '/auth/refresh-token',
      {
        refreshToken,
      }
    );

    return response.data;
  }

  // LOGOUT
  async logout(refreshToken: string) {
    const response = await apiClient.post(
      '/auth/logout',
      {
        refreshToken,
      }
    );

    return response.data;
  }
}

export default new AuthService();